### PR TITLE
methods-defining_methods-hard-q8

### PR DESCRIPTION
**Pull request by ketphan02**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”
---
## Website
- [x] Is the question text on the Gamification document up to date? 
    > Make sure to include the sample output if there is one for the question

### CODE REVIEW

File: src/methods/defining_methods/hard/q8/RentMethod.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package methods.defining_methods.hard.q8;
+
+/**
+ * [J] Write a Java Program that asks users for the cost of rent for their housing per month. Then
```
Comment (Hedgemon4) [inline src/methods/defining_methods/hard/q8/RentMethod.java:4]: Change question text to: 
`Write a Java Program that asks users for the cost of rent for their housing per month. Then write a Java method called rentCalc that accepts that value as a double and returns the cost per year for housing as a double. Your program should print the returned value in the main method. `

File: src/methods/defining_methods/hard/q8/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        Arguments.of(700.0, 8400.0),
+        Arguments.of(1000.0, 12000.0),
+        Arguments.of(321.344, 3856.128),
+        Arguments.of(1.1, 13.2));
+  }
+
+  @ParameterizedTest
+  @MethodSource("rentPricesProvider")
+  void correctCalculatesRent(double rentMonthly, double rentYearly) throws Throwable {
+    String errorMsg = "Your method \"rentCalcYear()\" does not calculate the yearly rent correctly.";
```
Comment (Hedgemon4) [inline src/methods/defining_methods/hard/q8/MainTest.java:54]: I would just change it so it has the method name here, like Your method "rentCalc"

File: src/methods/defining_methods/hard/q8/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    Object output = MethodUtil.invokeIfMethodExists(RentMethod.class, "rentCalcYear", new Object[] { rentMonthly },
+        double.class);
+    CustomAssertions._assertEquals(output, rentYearly, 0.01, errorMsg);
```
Comment (Hedgemon4) [inline src/methods/defining_methods/hard/q8/MainTest.java:57]: We also need to make sure here that we check and see if the student has printed any string output in the method, as they were instructed to print it in the main method

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/methods/defining_methods/hard/q8/RentMethod.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package methods.defining_methods.hard.q8;
+
+/*
+Write a Java Program that asks users for the cost of rent for their housing per month. Then write a Java method called
+rentCalcYear that accepts that value as a double and returns the cost per year for housing as a double. Your program should
+print the returned value in the main method.
```
Comment (Hedgemon4) [inline src/methods/defining_methods/hard/q8/RentMethod.java:6]: For this one add-in that the user will enter a rent greater than or equal to 0

File: src/methods/defining_methods/hard/q8/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    static Stream<Arguments> rentPricesProvider() {
+        return Stream.of(
+                Arguments.of(100.0, 1200.0),
+                Arguments.of(200.0, 2400.0),
+                Arguments.of(500.0, 6000.0),
+                Arguments.of(600.0, 7200.0),
+                Arguments.of(700.0, 8400.0),
+                Arguments.of(1000.0, 12000.0),
+                Arguments.of(321.344, 3856.128),
+                Arguments.of(1.1, 13.2));
```
Comment (Hedgemon4) [inline src/methods/defining_methods/hard/q8/MainTest.java:51]: Add a test for 0

File: src/methods/defining_methods/hard/q8/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+    public void runMain() {
+        RentMethod.main(new String[0]);
+    }
+
+    static Stream<Arguments> rentPricesProvider() {
+        return Stream.of(
+                Arguments.of(100.0, 1200.0),
+                Arguments.of(200.0, 2400.0),
+                Arguments.of(500.0, 6000.0),
```
Comment (Hedgemon4) [inline src/methods/defining_methods/hard/q8/MainTest.java:46]: Also maybe use more double values with numbers other than zero in the decimal (to make sure they are not losing precision.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/methods/defining_methods/hard/q8/RentMethod.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
+package methods.defining_methods.hard.q8;
+
+/*
+Write a Java Program that asks users for the cost of rent for their housing per month. Then write a Java method called
+rentCalcYear that accepts that value as a double and returns the cost per year for housing as a double. Your program should
+print the returned value in the main method. Assume a user input of 0 or greater.
```
Comment (Hedgemon4) [inline src/methods/defining_methods/hard/q8/RentMethod.java:6]:
```suggestion
print the returned value in the main method. Assume a user input of 0 or more.
```

File: src/methods/defining_methods/hard/q8/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+                Arguments.of(700.99, 8411.88),
+                Arguments.of(1000.0, 12000.0),
+                Arguments.of(321.344, 3856.128),
+                Arguments.of(1.1, 13.2),
+                Arguments.of(0, 0));
+    }
+
+    @ParameterizedTest
+    @MethodSource("rentPricesProvider")
+    void correctCalculatesRent(double rentMonthly, double rentYearly) throws Throwable {
```
Comment (Hedgemon4) [inline src/methods/defining_methods/hard/q8/MainTest.java:57]:
```suggestion
    void correctCalculatesRentMethod(double rentMonthly, double rentYearly) throws Throwable {
```

File: src/methods/defining_methods/hard/q8/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        String errorMsg = "Your rentCalc() method does not calculate the yearly rent correctly.";
+        CustomAssertions._assertEquals(output, rentYearly, 0.00001, errorMsg);
+    }
+
+    @ParameterizedTest
+    @MethodSource("rentPricesProvider")
+    void printsCorrectYearlyRent(double rentMonthly, double rentYearly) {
+        String errorMsg = "Your program does not print the correct yearly rent.";
+        runWithInput(rentMonthly + "");
+        assertEquals(Double.parseDouble(getItemByName("rentYearly")), rentYearly, 0.00001, errorMsg);
```
Comment (Hedgemon4) [inline src/methods/defining_methods/hard/q8/MainTest.java:72]: Should be rentYearly here first

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/methods/defining_methods/hard/q8/MainTest.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        return new Clause[]{
+                new StringLiteral("Enter the cost of rent per month:"),
+                new Optional(new StringLiteral(" ")),
+                new NewLine(),
+                new StringLiteral("The total cost of rent per year is \\$"),
+                new DoubleLiteral("rentYearly")
+        };
+    }
+
+    ;
```
[the remark this answers is not in the document]
Comment (SamamthaH) [reply src/methods/defining_methods/hard/q8/MainTest.java:36]: agreed

**Review by Hedgemon4 — APPROVED** [review]

**Pull request merged**